### PR TITLE
feat(ipkg): add parser and queries

### PR DESCRIPTION
**Pull request by srghma**

### CODE REVIEW

**Comment by srghma** [comment]
@clason finished, tnx )

**Comment by ribru17** [comment]
Seems like the parser is not very mature? Errors on a simple file:

```
package crypto

sourcedir = src
modules = Data.Crypto.Encryption.Block.DEA,
          Data.Crypto.Encryption.Block.EncryptionMode,
          Data.Crypto.Encryption.Block.TDEA,
          Data.Crypto.Encryption.Stream.ARC4,
          Data.Crypto.Hash.MD5
          -- Data.Crypto.Hash.SHS,
          -- Data.Crypto.MAC.HMAC

opts = "-p contrib"
```

**Comment by srghma** [comment]
Uncaught error: Error: Expected string.

"micropack.ipkg":23:14--23:17
 19 | main       = MicroPack
 20 |
 21 | executable = micropack
 22 |
 23 | sourcedir  = src
                   ^^^
`src` should be `"src"`

File: lua/nvim-treesitter/parsers.lua [hunk 1]
```diff
@@ -1017,6 +1017,14 @@ list.hyprlang = {
   maintainers = { "@luckasRanarison" },
 }
 
+list.ipkg = {
```
Comment (clason) [inline lua/nvim-treesitter/parsers.lua:1020]: Please keep these lists (parsers, lockfile, readme) alphabetical.

File: queries/ipkg/highlights.scm [hunk 2]
```diff
@@ -0,0 +1,52 @@
...
+  "version"
+  "langversion"
+  "modules"
+  "main"
+  "executable"
+  "depends"
+] @keyword
+
+; Package name
+(package_name) @type
```
Comment (clason) [inline queries/ipkg/highlights.scm:31]: Please group all same "simple" captures into one alternation (like `@keyword` above).

File: queries/ipkg/highlights.scm
Comment (ObserverOfTime) [inline queries/ipkg/highlights.scm:1]: The captures are not quite right. Check `CONTRIBUTING.md`.

Comment (srghma) [reply queries/ipkg/highlights.scm:1]: tnx, fixed?

Comment (ObserverOfTime) [reply queries/ipkg/highlights.scm:1]: No.

Comment (srghma) [reply queries/ipkg/highlights.scm:1]: Could You tell what mistakes You observe?

**Comment by srghma** [comment]
@clason lets merge?

File: queries/ipkg/highlights.scm
Comment (srghma) [reply queries/ipkg/highlights.scm:1]: @ObserverOfTime friendly ping

File: queries/ipkg/highlights.scm [hunk 3]
```diff
@@ -0,0 +1,46 @@
+; Keywords
+[
+  "package"
```
Comment (ObserverOfTime) [inline queries/ipkg/highlights.scm:3]: "package" should be `@keyword.import`

File: queries/ipkg/highlights.scm [hunk 4]
```diff
@@ -0,0 +1,46 @@
...
+  "modules"
+  "main"
+  "executable"
+  "depends"
+] @keyword
+
+[
+  "="
+  ","
+] @punctuation.delimiter
```
Comment (ObserverOfTime) [inline queries/ipkg/highlights.scm:33]:
```suggestion
"=" @operator

"," @punctuation.delimiter
```

File: queries/ipkg/highlights.scm [hunk 5]
```diff
@@ -0,0 +1,46 @@
...
+
+; Field values
+(string_value) @string
+
+(boolean_value) @constant.builtin
```
Comment (ObserverOfTime) [inline queries/ipkg/highlights.scm:38]:
```suggestion
(boolean_value) @boolean
```

File: queries/ipkg/highlights.scm [hunk 6]
```diff
@@ -0,0 +1,46 @@
...
+
+; Version numbers and ranges
+(version_number) @number
+
+(version_range_op) @keyword.operator
```
Comment (ObserverOfTime) [inline queries/ipkg/highlights.scm:43]:
```suggestion
(version_range_op) @operator
```

File: queries/ipkg/highlights.scm [hunk 7]
```diff
@@ -0,0 +1,46 @@
...
+
+; Comments
+(comment) @comment
```
Comment (ObserverOfTime) [inline queries/ipkg/highlights.scm:46]:
```suggestion
(comment) @comment @spell
```

**Review by ObserverOfTime — CHANGES_REQUESTED** [review]

File: queries/ipkg/highlights.scm (hunk 3 above)
Comment (srghma) [reply queries/ipkg/highlights.scm:3]: but "package" is a name of a currently created package

the `@keyword.import` is more like for `depends = package1, package2`

or `modules = Module1, Module2`

---------------

actually, forget, You are right, the color should be different

**Review by ObserverOfTime — APPROVED** [review]

**Comment by srghma** [comment]
lets merge?

**Comment by clason** [comment]
Note that this parser is useless without filetype support in Vim.

**Comment by srghma** [comment]
@clason what should I do then?

How people map them? I cannot find in https://github.com/AstroNvim/astrocommunity/blob/main/lua/astrocommunity/pack/haskell/init.lua

here? https://github.com/nvim-treesitter/nvim-treesitter/blob/01a92da4389b5346b4049c5b42256c9d84526f04/lua/nvim-treesitter/parsers.lua#L30

but before You told to delete a mapping there

**Comment by clason** [comment]
Make a PR to vim/vim.

**Comment by srghma** [comment]
@clason I have created a pr https://github.com/neovim/neovim/pull/31068

lets merge this pr?

**Comment by lucario387** [comment]
Filetype stuffs need PR in vim first, then neovim will backport it

**Comment by clason** [comment]
I will merge it when I believe it is ready to be merged; not sooner nor later.

File: queries/ipkg/highlights.scm [hunk 8]
```diff
@@ -26,7 +26,7 @@
   "main"
   "executable"
   "depends"
-] @keyword
+] @keyword.operator
```
Comment (clason) [inline queries/ipkg/highlights.scm:29]: No.

Comment (srghma) [reply queries/ipkg/highlights.scm]: reverted

File: queries/ipkg/highlights.scm [hunk 9]
```diff
@@ -0,0 +1,47 @@
...
+  "langversion"
+  "modules"
+  "main"
+  "executable"
+  "depends"
+] @keyword
+
+[
+  "="
+  (version_range_op)
```
Comment (clason) [inline queries/ipkg/highlights.scm:33]: I was talking about this.

**Comment by srghma** [comment]
dont know why it errors - I didnt rename "ipkg" to "ipkg2", just added scaner.c and `block_comment` `{- asdfasdf -}` parsing

https://github.com/srghma/tree-sitter-ipkg/commit/f46a3a02018620378eef5f3e128110d35c676331

https://github.com/srghma/tree-sitter-ipkg/commits/main/

**Review by ribru17 — APPROVED** [review]

**Comment by srghma** [comment]
@clason @ribru17 done

File: queries/ipkg/highlights.scm [hunk 10]
```diff
@@ -0,0 +1,56 @@
...
+  "version"
+  "langversion"
+  "modules"
+  "main"
+  "executable"
+  "depends"
+] @keyword
+
+[
+  "="
```
Comment (clason) [inline queries/ipkg/highlights.scm:32]: This should be `@operator`, not `@keyword.operator` (it's not an actual word).

**Review by ribru17 — APPROVED** [review]

**Pull request merged**